app/lib/account/backend.dart

```diff
@@ -0,0 +1,195 @@
+// Copyright (c) 2019, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:gcloud/db.dart';
+import 'package:gcloud/service_scope.dart' as ss;
+import 'package:googleapis/oauth2/v2.dart' as oauth2_v2;
+import 'package:http/http.dart' as http;
+import 'package:logging/logging.dart';
+import 'package:pool/pool.dart';
+import 'package:retry/retry.dart';
+
+import 'models.dart';
+
+final _logger = new Logger('pub.account.backend');
+final _retry = RetryOptions();
+
+/// The pub client's OAuth2 identifier.
+final _pubAudience = '818368855108-8grd2eg9tj9f38os6f1urbcvsq399u8n.apps.'
+    'googleusercontent.com';
+
+/// Sets the account backend service.
+void registerAccountBackend(AccountBackend backend) =>
+    ss.register(#_accountBackend, backend);
+
+/// The active account backend service.
+AccountBackend get accountBackend =>
+    ss.lookup(#_accountBackend) as AccountBackend;
+
+/// Represents the backend for the account handling and authentication.
+class AccountBackend {
+  final _pool = Pool(1);
+  final DatastoreDB _db;
+  final _defaultAuthProvider =
+      GoogleOauth2AuthProvider('dartlang-pub-1', _pubAudience);
+
+  AccountBackend(this._db);
+
+  Future close() async {
+    await _defaultAuthProvider.close();
+    await _pool.close();
+  }
+
+  /// Returns the `User` entry that is associated with the [accessToken].
+  ///
+  /// When no entry exists in Datastore, this method will create a new one.
+  ///
+  /// When the authenticated e-mail of the user changes, the email field will
+  /// be updated to the latest one, and the `historicalEmails` field contains
+  /// all of the earlier and current values.
+  ///
+  /// The method returns null if the access token is invalid.
+  Future<User> lookupOrCreateUser(String accessToken) async {
```

Comment:
How about we create two methods:
 * `authenticateWithAccessToken(String accessToken)`
 * `authenticateWithIdToken(String idToken)`

Or even better we just a take `authenticateUser(request)`, then we can add all the auth methods we want... using cookies or different bearer tokens, etc...